lib/ansible/modules/crypto/openssh_cert.py

```diff
@@ -0,0 +1,544 @@
+            if not dispatched_time:
+                module.fail_json(msg="'%s' is not a valid time format." % timestring)
+            dispatched_time = dispatched_time[0]
+            if dispatched_time[0] == "+":
+                return datetime.utcnow() + timedelta(
+                    weeks=int('0' + dispatched_time[2]),
+                    days=int('0' + dispatched_time[4]),
+                    hours=int('0' + dispatched_time[6]),
+                    minutes=int('0' + dispatched_time[8]),
+                    seconds=int('0' + dispatched_time[10]))
+            else:
+                return datetime.utcnow() - timedelta(
+                    weeks=int('0' + dispatched_time[2]),
+                    days=int('0' + dispatched_time[4]),
+                    hours=int('0' + dispatched_time[6]),
+                    minutes=int('0' + dispatched_time[8]),
+                    seconds=int('0' + dispatched_time[10]))
+        else:
+            formats = ["%Y:%m:%d", "%Y:%m:%d %H:%M:%S", "%Y:%m:%d:%H:%M:%S", "%Y-%m-%dT%H:%M:%S"]
+            for fmt in formats:
+                try:
+                    return datetime.strptime(timestring, fmt)
+                except:
+                    pass
+            module.fail_json(msg="'%s' is not a valid time format" % timestring)
+
+    def is_relative(self, timestr):
+        if timestr.startswith("+") or timestr.startswith("-"):
+            return True
+        return False
+
+    def is_valid(self, module, perms_required=True):
+
+        def _check_state():
+            return os.path.exists(self.path)
+
+        if _check_state():
+            proc = module.run_command([self.ssh_keygen, '-L', '-f', self.path], environ_update=dict(TZ="UTC"))
+            if proc[0] != 0:
+                return False
+            self.cert_info = proc[1].split()
+            principals = re.findall("(?<=Principals:)(.*)(?=Critical)", proc[1], re.S)[0].split()
+            principals = list(map(str.strip, principals))
+            cert_type = re.findall("( user | host )", proc[1])[0].strip()
+            validity = re.findall("(from (\\d{4}-\\d{2}-\\d{2}T\\d{2}(:\\d{2}){2}) to (\\d{4}-\\d{2}-\\d{2}T\\d{2}(:\\d{2}){2}))", proc[1])
+            if validity:
+                cert_valid_from = self.convert_to_datetime(module, validity[0][1])
+                cert_valid_to = self.convert_to_datetime(module, validity[0][3])
+            else:
+                cert_valid_from = datetime(MINYEAR, 1, 1)
+                cert_valid_to = datetime(MAXYEAR, 12, 31)
+        else:
+            return False
+
+        def _check_perms(module):
+            file_args = module.load_file_common_arguments(module.params)
+            return not module.set_fs_attributes_if_different(file_args, False)
+
+        def _check_type():
+            return self.type == cert_type
+
+        def _check_principals():
+            return self.principals == principals
```

Comment:
This check always fails on my machine: `self.principals == None`, `principals == ['(none)']`